src/construction/pk_encryption/ccs_from_ibe.rs

```diff
@@ -0,0 +1,150 @@
+/// public key encryption construction based on [\[5\]](<index.html#:~:text=[5]>).
+///
+/// Attributes:
+/// - `ibe`: specifies the IBE scheme used in this construction
+/// - `signature`: specifies the signature scheme used in this construction
+///
+/// # Examples
+/// ```
+/// use qfall_crypto::construction::pk_encryption::{CCSfromIBE, PKEncryptionMut};
+/// use qfall_math::integer::Z;
+/// let mut scheme = CCSfromIBE::init_dr_pfdh_from_n(4);
+///
+/// let (pk, sk) = scheme.gen();
+/// let cipher = scheme.enc(&pk, 0);
+/// let m = scheme.dec(&sk, &cipher);
+///
+/// assert_eq!(Z::ZERO, m);
+/// ```
+#[derive(Serialize, Deserialize, Clone)]
+pub struct CCSfromIBE<IBE: IdentityBasedEncryption, Signature: SignatureScheme>
+where
+    IBE::Cipher: ToString,
+{
+    pub ibe: IBE,
+    pub signature: Signature,
+}
+
+impl<IBE, Signature> PKEncryptionMut for CCSfromIBE<IBE, Signature>
+where
+    IBE: IdentityBasedEncryption,
+    Signature: SignatureScheme,
+    IBE::Cipher: ToString,
+    IBE::MasterPublicKey: Clone,
+    Signature::PublicKey: Into<IBE::Identity> + Clone,
+{
+    type Cipher = (Signature::PublicKey, IBE::Cipher, Signature::Signature);
+    type PublicKey = IBE::MasterPublicKey;
+    type SecretKey = (IBE::MasterPublicKey, IBE::MasterSecretKey);
+
+    /// Generates a (pk, sk) pair for the CCS construction
+    /// by following these steps:
+    /// - (mpk, msk) = ibe.setup()
+    ///
+    /// Then, `pk = mpk` and `sk = (mpk, msk)` are returned.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_crypto::construction::pk_encryption::{CCSfromIBE, PKEncryptionMut};
+    /// let mut scheme = CCSfromIBE::init_dr_pfdh_from_n(4);
+    ///
+    /// let (pk, sk) = scheme.gen();
+    /// ```
+    fn gen(&mut self) -> (Self::PublicKey, Self::SecretKey) {
+        let (pk, sk) = self.ibe.setup();
+        (pk.clone(), (pk, sk))
+    }
+
+    /// Generates an encryption of `message mod 2` for the provided public key by following these steps:
+    /// - (vrfy_key, sign_key) = signature.gen()
+    /// - c = ibe.enc(mpk, vrfy_key, message), i.e. encrypt `message mod 2` with respect to identity `vrfy_key`
+    /// - sigma = signature.sign(c, sign_key, vrfy_key), i.e. sign message `c`
+    ///
+    /// Then, the ciphertext `(vrfy_key, c, sigma)` is returned.
+    ///
+    /// Parameters:
+    /// - `pk`: specifies the public key `pk = A`
+    /// - `message`: specifies the message that should be encrypted
+    ///
+    /// Returns a cipher consisting of a tuple `cipher = (vrfy_key, c, sigma)`.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_crypto::construction::pk_encryption::{CCSfromIBE, PKEncryptionMut};
+    /// let mut scheme = CCSfromIBE::init_dr_pfdh_from_n(4);
+    ///
+    /// let (pk, sk) = scheme.gen();
+    /// let cipher = scheme.enc(&pk, 1);
+    /// ```
+    fn enc(&mut self, pk: &Self::PublicKey, message: impl Into<Z>) -> Self::Cipher {
+        let message = message.into().modulo(2);
```

Comment:
if the encryption scheme allows for the encryption of larger messages, this would be heavily limit the message space for the encryption scheme

Reply:
I removed the modulus application as I thought one step too far at that point (and just thought about the DualRegev encryption), but of course some generic IBE could be able to encrypt more than one bit at a time.